pgjdbc/src/main/java/org/postgresql/largeobject/BlobInputStream.java

```diff
@@ -76,43 +77,131 @@ public BlobInputStream(LargeObject lo, int bsize) {
    */
   public BlobInputStream(LargeObject lo, int bsize, long limit) {
     this.lo = lo;
-    buffer = null;
-    bpos = 0;
-    apos = 0;
-    this.bsize = bsize;
-    this.limit = limit;
+    this.bufferSize = bsize;
+    // Treat -1 as no limit for backward compatibility
+    this.limit = limit == -1 ? Long.MAX_VALUE : limit;
   }
 
   /**
    * The minimum required to implement input stream.
    */
   public int read() throws java.io.IOException {
-    LargeObject lo = getLo();
-    try {
-      if (limit > 0 && apos >= limit) {
+    try (ResourceLock ignore = lock.obtain()) {
+      LargeObject lo = getLo();
```

Comment:
I'd probably move this down below the check for limit, but lo isn't an expensive operation so I'm +/- 0

Reply:
I guess we do not want to optimize "somebody calling read many times at the end of the stream".

At the same time, `lo` value might change if somebody closes the lob in a different thread (or even calls `read()`), so I would rather synchronize the methods

Reply:
If you attempt `read` on a closed stream, you probably want `BlobOutputStream is closed` rather than `-1`.

Reply:
So throw an exception ?